Dockerfile

```diff
@@ -1,13 +1,33 @@
-FROM python:3.7
+FROM node:12 as frontend
 
-ENV FLASK_APP=dashboard:create_app
+RUN mkdir /tmp/frontend
+WORKDIR /tmp/frontend
+
+# cache hack; fragile
+COPY package.json package-lock.json ./
+RUN npm install
+
+COPY . .
+RUN npm run build
+
+# -----------------------------------------------------------------------------
+FROM python:3.7 as backend
 
-RUN mkdir /code
-WORKDIR /code
+RUN mkdir /opt/stayhome-dashboard
+WORKDIR /opt/stayhome-dashboard
+
+# todo: reduce to single (sub)directory for easier copying
+# Copy front-end files built in previous stage
+COPY --from=frontend /tmp/frontend/dashboard/static/js/ /opt/stayhome-dashboard/dashboard/static/js/
+COPY --from=frontend /tmp/frontend/dashboard/templates/ /opt/stayhome-dashboard/dashboard/templates/
+
```

Comment:
Ivan, I added a new directory, `public`, to house things e.g. json, favicon, that won't/can't be auto-generated, see my PR:
https://github.com/uwcirg/stayhome-dashboard/pull/5
So we will probably need to do this?  `COPY --from=frontend /tmp/frontend/dashboard/static/public/ /opt/stayhome-dashboard/dashboard/static/public/`  in the event that PR gets merged.  Otherwise I think the frontend portion looks good to me.

Reply:
Thanks for the heads up! I'll merge that first and fix this afterwards.

All I should have to do is `COPY` the whole `static/` directory instead of just `static/js/`.

Reply:
@achen2401 
I'm getting the following build error after merging develop into my branch (after merging your PR into develop). Could you take a look when you get a chance?

```
Version: webpack 4.28.0
Time: 5317ms
Built at: 04/03/2020 4:15:29 PM
 10 assets
Entrypoint main = 2.app.bundle.js 1.app.bundle.js 7.app.bundle.js 3.app.bundle.js 8.app.bundle.js 5.app.bundle.js 6.app.bundle.js 4.app.bundle.js 9.app.bundle.js app.bundle.js
 [6] ./node_modules/@material-ui/core/esm/styles/withStyles.js + 2 modules 5.31 KiB {2} [built]
     |    3 modules
[13] ./node_modules/@babel/runtime/helpers/esm/toConsumableArray.js + 3 modules 914 bytes {1} [built]
     |    4 modules
[18] ./node_modules/@material-ui/core/esm/utils/useForkRef.js + 1 modules 749 bytes {2} [built]
     |    2 modules
[27] ./node_modules/@material-ui/core/esm/styles/defaultTheme.js + 18 modules 30.5 KiB {2} [built]
     |    19 modules
[36] ./node_modules/react-table-6/es/index.js + 6 modules 307 KiB {7} [built]
     |    7 modules
[37] ./node_modules/@babel/runtime/helpers/esm/slicedToArray.js + 3 modules 1.25 KiB {1} [built]
     |    4 modules
[54] ./dashboard/static/src/less/index.less 639 bytes {0} [built]
[56] ./node_modules/css-loader/dist/cjs.js!./node_modules/less-loader/dist/cjs.js??ref--6-2!./dashboard/static/src/less/index.less 4.56 KiB {0} [built]
[58] ./dashboard/static/src/js/Index.js + 6 modules 9.69 KiB {0} [built]
     | ./dashboard/static/src/js/Index.js 186 bytes [built]
     | ./dashboard/static/src/js/App.js 584 bytes [built]
     | ./dashboard/static/src/assets/img/stayhome-logo.png 1.75 KiB [built]
     |     + 4 hidden modules
[85] ./node_modules/@material-ui/core/esm/Box/Box.js + 14 modules 15.4 KiB {2} [built]
     |    15 modules
[86] ./node_modules/@material-ui/styles/esm/makeStyles/makeStyles.js + 9 modules 20.4 KiB {2} [built]
     |    10 modules
[87] ./node_modules/@material-ui/core/esm/Button/Button.js + 5 modules 42.3 KiB {2} [built]
     |    6 modules
[89] ./node_modules/@material-ui/core/esm/AppBar/AppBar.js + 1 modules 8.07 KiB {2} [built]
     |    2 modules
[90] ./node_modules/@material-ui/system/esm/spacing.js + 1 modules 4.21 KiB {2} [built]
     |    2 modules
[91] ./node_modules/@material-ui/styles/esm/useTheme/useTheme.js + 1 modules 349 bytes {2} [built]
     |    2 modules
    + 104 hidden modules

ERROR in ./node_modules/react-table-6/react-table.css 1:0
Module parse failed: Unexpected token (1:0)
You may need an appropriate loader to handle this file type.
> .ReactTable{position:relative;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;border:1px solid rgba(0,0,0,0.1);}.ReactTable *{box-sizing:border-box}.ReactTable .rt-table{-webkit-box-flex:1;-ms-flex:auto 1;flex:auto 1;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-box-align:stretch;-ms-flex-align:stretch;align-items:stretch;width:100%;border-collapse:collapse;overflow:auto}.ReactTable .rt-thead{-webkit-box-flex:1;-ms-flex:1 0 auto;flex:1 0 auto;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;}.ReactTable .rt-thead.-headerGroups{background:rgba(0,0,0,0.03);border-bottom:1px solid rgba(0,0,0,0.05)}.ReactTable .rt-thead.-filters{border-bottom:1px solid rgba(0,0,0,0.05);}.ReactTable .rt-thead.-filters input,.ReactTable .rt-thead.-filters select{border:1px solid rgba(0,0,0,0.1);background:#fff;padding:5px 7px;font-size:inherit;border-radius:3px;font-weight:normal;outline-width:0}.ReactTable .rt-thead.-filters .rt-th{border-right:1px solid rgba(0,0,0,0.02)}.ReactTable .rt-thead.-header{box-shadow:0 2px 15px 0 rgba(0,0,0,0.15)}.ReactTable .rt-thead .rt-tr{text-align:center}.ReactTable .rt-thead .rt-th,.ReactTable .rt-thead .rt-td{padding:5px 5px;line-height:normal;position:relative;border-right:1px solid rgba(0,0,0,0.05);transition:box-shadow .3s cubic-bezier(.175,.885,.32,1.275);box-shadow:inset 0 0 0 0 transparent;}.ReactTable .rt-thead .rt-th.-sort-asc,.ReactTable .rt-thead .rt-td.-sort-asc{box-shadow:inset 0 3px 0 0 rgba(0,0,0,0.6)}.ReactTable .rt-thead .rt-th.-sort-desc,.ReactTable .rt-thead .rt-td.-sort-desc{box-shadow:inset 0 -3px 0 0 rgba(0,0,0,0.6)}.ReactTable .rt-thead .rt-th.-cursor-pointer,.ReactTable .rt-thead .rt-td.-cursor-pointer{cursor:pointer}.ReactTable .rt-thead .rt-th:last-child,.ReactTable .rt-thead .rt-td:last-child{border-right:0}.ReactTable .rt-thead .rt-th:focus{outline-width:0}.ReactTable .rt-thead .rt-resizable-header{overflow:visible;}.ReactTable .rt-thead .rt-resizable-header:last-child{overflow:hidden}.ReactTable .rt-thead .rt-resizable-header-content{overflow:hidden;text-overflow:ellipsis}.ReactTable .rt-thead .rt-header-pivot{border-right-color:#f7f7f7}.ReactTable .rt-thead .rt-header-pivot:after,.ReactTable .rt-thead .rt-header-pivot:before{left:100%;top:50%;border:solid transparent;content:" ";height:0;width:0;position:absolute;pointer-events:none}.ReactTable .rt-thead .rt-header-pivot:after{border-color:rgba(255,255,255,0);border-left-color:#fff;border-width:8px;margin-top:-8px}.ReactTable .rt-thead .rt-header-pivot:before{border-color:rgba(102,102,102,0);border-left-color:#f7f7f7;border-width:10px;margin-top:-10px}.ReactTable .rt-tbody{-webkit-box-flex:99999;-ms-flex:99999 1 auto;flex:99999 1 auto;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;overflow:auto;}.ReactTable .rt-tbody .rt-tr-group{border-bottom:solid 1px rgba(0,0,0,0.05);}.ReactTable .rt-tbody .rt-tr-group:last-child{border-bottom:0}.ReactTable .rt-tbody .rt-td{border-right:1px solid rgba(0,0,0,0.02);}.ReactTable .rt-tbody .rt-td:last-child{border-right:0}.ReactTable .rt-tbody .rt-expandable{cursor:pointer;text-overflow:clip}.ReactTable .rt-tr-group{-webkit-box-flex:1;-ms-flex:1 0 auto;flex:1 0 auto;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;-webkit-box-align:stretch;-ms-flex-align:stretch;align-items:stretch}.ReactTable .rt-tr{-webkit-box-flex:1;-ms-flex:1 0 auto;flex:1 0 auto;display:-webkit-inline-box;display:-ms-inline-flexbox;display:inline-flex}.ReactTable .rt-th,.ReactTable .rt-td{-webkit-box-flex:1;-ms-flex:1 0 0px;flex:1 0 0;white-space:nowrap;text-overflow:ellipsis;padding:7px 5px;overflow:hidden;transition:.3s ease;transition-property:width,min-width,padding,opacity;}.ReactTable .rt-th.-hidden,.ReactTable .rt-td.-hidden{width:0 !important;min-width:0 !important;padding:0 !important;border:0 !important;opacity:0 !important}.ReactTable .rt-expander{display:inline-block;position:relative;margin:0;color:transparent;margin:0 10px;}.ReactTable .rt-expander:after{content:'';position:absolute;width:0;height:0;top:50%;left:50%;-webkit-transform:translate(-50%,-50%) rotate(-90deg);transform:translate(-50%,-50%) rotate(-90deg);border-left:5.04px solid transparent;border-right:5.04px solid transparent;border-top:7px solid rgba(0,0,0,0.8);transition:all .3s cubic-bezier(.175,.885,.32,1.275);cursor:pointer}.ReactTable .rt-expander.-open:after{-webkit-transform:translate(-50%,-50%) rotate(0);transform:translate(-50%,-50%) rotate(0)}.ReactTable .rt-resizer{display:inline-block;position:absolute;width:36px;top:0;bottom:0;right:-18px;cursor:col-resize;z-index:10}.ReactTable .rt-tfoot{-webkit-box-flex:1;-ms-flex:1 0 auto;flex:1 0 auto;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;box-shadow:0 0 15px 0 rgba(0,0,0,0.15);}.ReactTable .rt-tfoot .rt-td{border-right:1px solid rgba(0,0,0,0.05);}.ReactTable .rt-tfoot .rt-td:last-child{border-right:0}.ReactTable.-striped .rt-tr.-odd{background:rgba(0,0,0,0.03)}.ReactTable.-highlight .rt-tbody .rt-tr:not(.-padRow):hover{background:rgba(0,0,0,0.05)}.ReactTable .-pagination{z-index:1;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-pack:justify;-ms-flex-pack:justify;justify-content:space-between;-webkit-box-align:stretch;-ms-flex-align:stretch;align-items:stretch;-ms-flex-wrap:wrap;flex-wrap:wrap;padding:3px;box-shadow:0 0 15px 0 rgba(0,0,0,0.1);border-top:2px solid rgba(0,0,0,0.1);}.ReactTable .-pagination input,.ReactTable .-pagination select{border:1px solid rgba(0,0,0,0.1);background:#fff;padding:5px 7px;font-size:inherit;border-radius:3px;font-weight:normal;outline-width:0}.ReactTable .-pagination .-btn{-webkit-appearance:none;-moz-appearance:none;appearance:none;display:block;width:100%;height:100%;border:0;border-radius:3px;padding:6px;font-size:1em;color:rgba(0,0,0,0.6);background:rgba(0,0,0,0.1);transition:all .1s ease;cursor:pointer;outline-width:0;}.ReactTable .-pagination .-btn[disabled]{opacity:.5;cursor:default}.ReactTable .-pagination .-btn:not([disabled]):hover{background:rgba(0,0,0,0.3);color:#fff}.ReactTable .-pagination .-previous,.ReactTable .-pagination .-next{-webkit-box-flex:1;-ms-flex:1;flex:1;text-align:center}.ReactTable .-pagination .-center{-webkit-box-flex:1.5;-ms-flex:1.5;flex:1.5;text-align:center;margin-bottom:0;display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;-ms-flex-direction:row;flex-direction:row;-ms-flex-wrap:wrap;flex-wrap:wrap;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-ms-flex-pack:distribute;justify-content:space-around}.ReactTable .-pagination .-pageInfo{display:inline-block;margin:3px 10px;white-space:nowrap}.ReactTable .-pagination .-pageJump{display:inline-block;}.ReactTable .-pagination .-pageJump input{width:70px;text-align:center}.ReactTable .-pagination .-pageSizeOptions{margin:3px 10px}.ReactTable .rt-noData{display:block;position:absolute;left:50%;top:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);background:rgba(255,255,255,0.8);transition:all .3s ease;z-index:1;pointer-events:none;padding:20px;color:rgba(0,0,0,0.5)}.ReactTable .-loading{display:block;position:absolute;left:0;right:0;top:0;bottom:0;background:rgba(255,255,255,0.8);transition:all .3s ease;z-index:-1;opacity:0;pointer-events:none;}.ReactTable .-loading > div{position:absolute;display:block;text-align:center;width:100%;top:50%;left:0;font-size:15px;color:rgba(0,0,0,0.6);-webkit-transform:translateY(-52%);transform:translateY(-52%);transition:all .3s cubic-bezier(.25,.46,.45,.94)}.ReactTable .-loading.-active{opacity:1;z-index:2;pointer-events:all;}.ReactTable .-loading.-active > div{-webkit-transform:translateY(50%);transform:translateY(50%)}.ReactTable .rt-resizing .rt-th,.ReactTable .rt-resizing .rt-td{transition:none !important;cursor:col-resize;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}
 @ ./dashboard/static/src/js/components/Table.js 3:0-39
 @ ./dashboard/static/src/js/App.js
 @ ./dashboard/static/src/js/Index.js
Child HtmlWebpackCompiler:
     1 asset
    Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
    [0] ./node_modules/html-webpack-plugin/lib/loader.js!./dashboard/static/src/index.html 906 bytes {0} [built]
[91mnpm[0m[91m ERR! code ELIFECYCLE
npm ERR! errno 2
[0m[91mnpm ERR! stayhome_dashboard@1.0.0 build: `webpack -p --progress --config webpack.config.js`
npm [0m[91mERR![0m[91m Exit status 2
[0m[91mnpm ERR! 
[0m[91mnpm[0m[91m [0m[91mERR![0m[91m Failed at the stayhome_dashboard@1.0.0 build script.
[0m[91mnpm ERR![0m[91m This is probably not a problem with npm. There is likely additional logging output above.
[0m[91m
[0m[91mnpm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-04-03T16_15_29_483Z-debug.log
[0mService 'web' failed to build: The command '/bin/sh -c npm run build' returned a non-zero code: 2

```

Reply:
@ivan-c Ivan, hmm, I didn't get an error when I tried, I removed my node_modules folder and did a fresh install `npm install` and then `npm run build`.   Did you try that?  or are you still getting errors?

Reply:
@achen2401 
I've only tried building with docker, so I don't have any of the typical build artifacts (`node_modules`, `templates/index.html`).

I also tried building with `docker-compose build --no-cache` to be sure nothing was cached between builds.

Would you mind trying to build my branch with `docker-compose build --no-cache`?

Also, could [this change](https://github.com/uwcirg/stayhome-dashboard/commit/b20f168d3a3dd82e50e499f07999f98522169690#diff-8890ffa833ef7f5d475263f389ae4029L13) be related?

Reply:
Fixed!